operators/endpointmetrics/pkg/util/status.go

```diff
@@ -10,45 +10,92 @@ import (
-		"Degraded": {
-			"type":    "Degraded",
-			"reason":  "Degraded",
-			"message": "Metrics collector deployment not successful"},
-		"NotSupported": {
-			"type":    "NotSupported",
-			"reason":  "NotSupported",
-			"message": "No Prometheus service found in this cluster"},
+	conditions = map[StatusConditionName]*oav1beta1.StatusCondition{
+		DeployedStatus: {
+			Type:    "Progressing",
+			Reason:  "Deployed",
+			Message: "Metrics collector deployed",
+			Status:  metav1.ConditionTrue,
+		},
+		DisabledStatus: {
+			Type:    "Disabled",
+			Reason:  "Disabled",
+			Message: "enableMetrics is set to False",
+			Status:  metav1.ConditionTrue,
+		},
+		DegradedStatus: {
+			Type:    "Degraded",
+			Reason:  "Degraded",
+			Message: "Metrics collector deployment not successful",
+			Status:  metav1.ConditionTrue,
+		},
+		NotSupportedStatus: {
+			Type:    "NotSupported",
+			Reason:  "NotSupported",
+			Message: "No Prometheus service found in this cluster",
+			Status:  metav1.ConditionTrue,
+		},
 	}
 )
 
-func ReportStatus(ctx context.Context, client client.Client, i *oav1beta1.ObservabilityAddon, t string, reportStatus bool) {
-	if !reportStatus {
-		return
-	}
-	i.Status.Conditions = []oav1beta1.StatusCondition{
-		{
-			Type:               conditions[t]["type"],
-			Status:             metav1.ConditionTrue,
-			LastTransitionTime: metav1.NewTime(time.Now()),
-			Reason:             conditions[t]["reason"],
-			Message:            conditions[t]["message"],
-		},
+func ReportStatus(ctx context.Context, client client.Client, condition StatusConditionName, addonName, addonNs string) error {
+	newCondition := conditions[condition].DeepCopy()
+	newCondition.LastTransitionTime = metav1.NewTime(time.Now())
+
+	// Fetch the ObservabilityAddon instance in local cluster, and update the status
+	// Retry on conflict
+	obsAddon := &oav1beta1.ObservabilityAddon{}
+	retryErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := client.Get(ctx, types.NamespacedName{Name: addonName, Namespace: addonNs}, obsAddon); err != nil {
+			return err
+		}
+
+		if !shouldAppendCondition(obsAddon.Status.Conditions, newCondition) {
+			return nil
+		}
+
+		obsAddon.Status.Conditions = append(obsAddon.Status.Conditions, *newCondition)
+
+		if len(obsAddon.Status.Conditions) > MaxStatusConditionsCount {
+			obsAddon.Status.Conditions = obsAddon.Status.Conditions[len(obsAddon.Status.Conditions)-MaxStatusConditionsCount:]
+		}
+
+		return client.Status().Update(ctx, obsAddon)
+	})
+	if retryErr != nil {
+		return retryErr
 	}
-	err := client.Status().Update(ctx, i)
-	if err != nil {
-		log.Error(err, "Failed to update status for observabilityaddon")
+
+	return nil
+}
+
+// shouldAppendCondition checks if the new condition should be appended to the status conditions
+// based on the last condition in the slice.
+func shouldAppendCondition(conditions []oav1beta1.StatusCondition, newCondition *oav1beta1.StatusCondition) bool {
```

Comment:
So it stops from changing the condition, if the condition is the same?

Reply:
The idea is to avoid adding duplicated status. If it is available, I don't need to report available again. 
But I assume here that statuses are sorted by date, which is not necessarily the case IIRC. Probably have to change this! Having a look.

Reply:
updated.